src/core/layout/qgslayoutexporter.cpp

```diff
@@ -43,26 +57,693 @@ void QgsLayoutExporter::renderPage( QPainter *painter, int page )
   renderRegion( painter, paperRect );
 }
 
-void QgsLayoutExporter::renderRegion( QPainter *painter, const QRectF &region )
+QImage QgsLayoutExporter::renderPageToImage( int page, QSize imageSize, double dpi ) const
+{
+  if ( !mLayout )
+    return QImage();
+
+  if ( mLayout->pageCollection()->pageCount() <= page || page < 0 )
+  {
+    return QImage();
+  }
+
+  QgsLayoutItemPage *pageItem = mLayout->pageCollection()->page( page );
+  if ( !pageItem )
+  {
+    return QImage();
+  }
+
+  QRectF paperRect = QRectF( pageItem->pos().x(), pageItem->pos().y(), pageItem->rect().width(), pageItem->rect().height() );
+  return renderRegionToImage( paperRect, imageSize, dpi );
+}
+
+///@cond PRIVATE
+class LayoutItemCacheSettingRestorer
+{
+  public:
+
+    LayoutItemCacheSettingRestorer( QgsLayout *layout )
+      : mLayout( layout )
+    {
+      mLayout->context().mIsPreviewRender = false;
+
+      const QList< QGraphicsItem * > items = mLayout->items();
+      for ( QGraphicsItem *item : items )
+      {
+        mPrevCacheMode.insert( item, item->cacheMode() );
+        item->setCacheMode( QGraphicsItem::NoCache );
+      }
+    }
+
+    ~LayoutItemCacheSettingRestorer()
+    {
+      for ( auto it = mPrevCacheMode.constBegin(); it != mPrevCacheMode.constEnd(); ++it )
```

Comment:
I don't see a lot of `auto` in QGIS source code. Do we have some kind of rules about its usage?

Reply:
Not officially (but we should open a discussion on dev list). From scattered conversation my summary is:

- don't use auto for declaring most variables. Explicit type declaration is preferred, as it helps readability and also qt creator code completion.
- let's use it only for annoying declarations, like `QMap< blah, blah >::const_iterator` 

I'm happy with this approach... personally I hate having to declare iterators, so always use auto for them now.

Reply:
>  - don't use auto for declaring most variables. Explicit type declaration is preferred, as it helps readability and also qt creator code completion.
> -  let's use it only for annoying declarations, like QMap< blah, blah >::const_iterator

Fully agree :+1: